lib/ecto/query/builder/windows.ex

```diff
@@ -0,0 +1,46 @@
+import Kernel, except: [apply: 2]
+
+defmodule Ecto.Query.Builder.Windows do
+  @moduledoc false
+
+  alias Ecto.Query.Builder
+  alias Ecto.Query.Builder.PartitionBy
+
+  @doc """
+  Builds a quoted expression.
+
+  The quoted expression should evaluate to a query at runtime.
+  If possible, it does all calculations at compile time to avoid
+  runtime work.
+  """
+  @spec build(Macro.t, [Macro.t], Keyword.t, Macro.Env.t) :: Macro.t
+  def build(query, binding, windows, env) do
+    {_, binding} = Builder.escape_binding(query, binding, env)
+    windows = windows
+              |> Enum.map(fn {name, expr} -> {name, PartitionBy.build(binding, expr, env)} end)
+
+    Builder.apply_query(query, __MODULE__, [windows], env)
+  end
+
+  @spec apply(Ecto.Queryable.t, Tuple.t) :: Ecto.Query.t
+  def apply_window({name, window}, %Ecto.Query{windows: windows} = query) do
+    if Keyword.has_key?(windows, name) do
+      Builder.error! "window with name #{name} is already defined"
+    else
+      %{query | windows: windows ++ [{name, window}]}
+    end
+  end
+
+  @doc """
+  The callback applied by `build/4` to build the query.
+  """
+  @spec apply(Ecto.Queryable.t, Keyword.t) :: Ecto.Query.t
+  def apply(%Ecto.Query{} = query, definitions) do
+    definitions
+    |> Enum.reduce(query, &apply_window/2)
```

Comment:
I would prefer something like:

```elixir
def apply(%{windows: windows} = query, definitions) do
  %{query | windows: windows ++ Enum.map(definitions, &validate_window!(&1, windows))}
end
```

Otherwise we are accessing the query and prepending the list of windows multiple times.